src/ast_editor.ml

```diff
@@ -0,0 +1,487 @@
+      None
+  in
+  match int_prop "selectedOutput" with
+  | Some i -> Ast_editor_state.set_original_mode ast_editor_state (i = 0)
+  | None -> (
+    match Option.both (int_prop "begin") (int_prop "end") with
+    | None -> ()
+    | Some (cbegin, cend) ->
+      let apply_selection editor cbegin cend =
+        let document = TextEditor.document editor in
+        let anchor = Vscode.TextDocument.positionAt document ~offset:cbegin in
+        let active = Vscode.TextDocument.positionAt document ~offset:cend in
+        TextEditor.set_selection editor
+          (Selection.makePositions ~anchor ~active);
+        TextEditor.revealRange editor
+          ~range:(Range.makePositions ~start:anchor ~end_:active)
+          ()
+      in
+      Vscode.Window.visibleTextEditors ()
+      |> List.iter ~f:(fun editor ->
+             let visible_doc = TextEditor.document editor in
+             if (* !original_mode && *) document_eq document visible_doc then
+               apply_selection editor cbegin cend
+             else if
+               (* (not !original_mode) && *)
+               (Ast_editor_state.entry_exists ast_editor_state
+                  ~origin_doc:(doc_string_uri ~document))
+                 ~pp_doc:(doc_string_uri ~document:visible_doc)
+               && not (Ast_editor_state.get_original_mode ast_editor_state)
+             then
+               match Option.both (int_prop "r_begin") (int_prop "r_end") with
+               | None -> ()
+               | Some (rcbegin, rcend) -> apply_selection editor rcbegin rcend))
+
+let on_hover custom_doc webview =
+  let provider =
+    let provideHover ~(document : TextDocument.t) ~(position : Position.t)
+        ~token:_ =
+      let offset = TextDocument.offsetAt document ~position in
+      if document_eq custom_doc document then
+        send_msg "focus" (Ojs.int_to_js offset) ~webview;
+      let hover =
+        Hover.make
+          ~contents:(`MarkdownString (MarkdownString.make ~value:"" ()))
+      in
+      `Value (Some hover)
+    in
+    HoverProvider.create ~provideHover
+  in
+  Vscode.Languages.registerHoverProvider ~selector:(`String "ocaml") ~provider
+
+let activate_hover_mode instance ~document =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match Ast_editor_state.find_webview_by_doc ast_editor_state ~document with
+  | Some webview -> on_hover document webview
+  | None -> failwith "Webview wasn't found while switching hover mode"
+
+let resolveCustomTextEditor instance ~(document : TextDocument.t) ~webviewPanel
+    ~token:_ : CustomTextEditorProvider.ResolvedEditor.t =
+  let webview = WebviewPanel.webview webviewPanel in
+  (*persist the webview*)
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  Ast_editor_state.set_webview_map ast_editor_state
+    (Map.set
+       (Ast_editor_state.get_webview_map ast_editor_state)
+       ~key:(doc_string_uri ~document) ~data:webview);
+  let (_ : Disposable.t) =
```

Comment:
Why is it ok to ignore this disposable? Shouldn't it be disposed when the plugin is disposed of?

Reply:
Actually I'm not sure what's the logic behind transforming the
```typescript
Event<void>
```
https://github.com/microsoft/vscode/blob/c3f126316369cd610563c75b1b1725e0679adfb3/src/vs/vscode.d.ts#L7608

to 
```OCaml 
Disposable.t 
```
But if we follow the similar [example](https://github.com/ocamllabs/vscode-ocaml-platform/blob/175a87213f9b024258dcffd4d7f6a6c3e89cbc59/src/repl.ml#L184) in `Repl` module, then we shouldn't ignore it. I did it just in case 6993d2647ebd03c9416e1086e0d593e02155a42f